### DISCUSSION TITLE
Upgrading from 9.2.0.1 to 9.5.0.2: How to get the size back under 512K?

### DISCUSSION THREAD

**Discussion by leres** [arendst/Tasmota#12562] [Q&A] [answered]
I'm running 9.2.0.1 on a rag tag assortment of devices including sonoff S31 and basic modules and a couple of 8266 dev boards. I use USE_MQTT_TLS and USE_MQTT_TLS_CA_CERT and have a config ([my_user_config.h](https://github.com/arendst/Tasmota/files/6760560/my_user_config.h.txt)) that fit in less that 512K which allowed OTA updates.

I just tried  upgrading to 9.5.0.1 and the bin was ~520K... I looked at my_user_config.h and don't see anything else I can turn off. What am I doing wrong?

**Comment by sfromis** [comment, marked as answer]
Well, always use the compressed versions (.bin.gz) for OTA upgrading. If you make a local web server, you can store the tasmota-minimal.bin.gz in the same folder, and Tasmota will automatically do a two-step upgrading (which you can also do manually).

**Reply by leres** [reply]
I'm already using compressed binaries with OTA updates, it's not **that** size that's the problem. It's my understanding that if I the "program size" reported by the "/in" page is more than 512K, there's not enough space in flash to store the  uncompressed program. I forgot to make a screen grab before I used esptool to revert to 9.2 but here's one that shows my current program size.

![tasmota](https://user-images.githubusercontent.com/597247/124399626-b006cf80-dcc0-11eb-8bfe-ba1f655d6d1d.png)

**Reply by sfromis** [reply]
It does not need to store both uncompressed binaries at the same time when OTA upgrading with a .gz binary, it can decompress in place.

**Reply by barbudor** [reply]
As sfromis said, if you use "File upload" from the web gui, then you have to manually go through `tasmota-minimal.bin.gz` first
Then upload `tasmota.bin.gz` (which is 434k, not 520k)

**Comment by Jason2866** [comment]
For converting devices from a different firmware which needs a build below 512k there is versiin Tasmota-lite.
